### PR TITLE
java: generated code to have javadoc

### PR DESCRIPTION
**Pull request by asashour**
**Merge with opencv_contrib**: https://github.com/opencv/opencv_contrib/pull/2131

Related to #4286

### This pullrequest changes

Adds javadoc to generated code. However, there are some current limitations:

- `@code` is copied as is.  I guess it needs to be decided if `@code{.java}` should exist in `.cpp` or in `gen_dict.json`, to be discussed in a new issue.
- Many doxygen commands are not properly handled, they should be fixed, in future PRs.
- `@param` is assumed to span a single line, not on multiple lines
- Embedded `images` are not handled
- Tests of the javadoc to be added (optionally, can we invoke `gen_java.py` with test code?)

To implement in this PR:
- [x] Formulas to be shown
- [X] Handle lists
- [X] simple replacements e.g. **\b**, `**bold**`, **\p**
- [x] Handle paragraphs
- [X] Add references to external Java doc (e.g. `java.util.List`)

```
opencv_contrib=param
force_builders=Docs
```

### CODE REVIEW

**Comment by alalek** [comment]
Javadoc generation [has been failed](http://pullrequest.opencv.org/buildbot/builders/precommit_docs/builds/20593/steps/make%20doxygen/logs/stdio) (whole build is green - `BUILD SUCCESSFUL` by Ant tool =( )

**Comment by asashour** [comment]
I need help.

In in the latest [build](https://pullrequest.opencv.org/buildbot/builders/precommit_docs/builds/20615/steps/make%20doxygen/logs/stdio), I don't know how `Bgsegm` is generated. It is not generated locally, even if `BUILD_DOCS` and `BUILD_EXAMPLES` are set to `ON`

Thanks a lot.

**Comment by alalek** [comment]
It is generated from https://github.com/opencv/opencv_contrib
```
cmake -DOPENCV_EXTRA_MODULES_PATH=<opencv_contrib>/modules ...
```

**Comment by asashour** [comment]
Thanks for support.

Just a hint, the build is green, but it seems not updated in this issue. Not sure if something should be done in `opencv_contrib` part, or `WIP` should be removed.

**Comment by alalek** [comment]
"status" waiting for all required builders. We will trigger them all before merge. Currently it is enough to launch "Docs" builder only.

Could you take a look on produced documentation?
Could you add more paragraphs, for example [here](http://pullrequest.opencv.org/buildbot/export/pr/14667/docs/javadoc/org/opencv/calib3d/Calib3d.html#solvePnP-org.opencv.core.MatOfPoint3f-org.opencv.core.MatOfPoint2f-org.opencv.core.Mat-org.opencv.core.MatOfDouble-org.opencv.core.Mat-org.opencv.core.Mat-)? (this link contents are updated automatically after "Docs" build completes).

**Comment by asashour** [comment]
Please also see updated description.

File: modules/java/jar/build.xml.in [hunk 1]
```diff
@@ -29,14 +29,33 @@
...
+      link="https://docs.oracle.com/javase/6/docs/api/"
+      additionalparam="--allow-script-in-comments"
+      >
+      <Header>
+         <![CDATA[
+            <script>
+              var url = window.location.href;
+              url = url.substring(0, url.lastIndexOf('/', url.indexOf('.html'))) + '/../../../mymath.js';
+              var script = document.createElement('script');
+              script.src = 'https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML,' + url;
```
Comment (alalek) [inline modules/java/jar/build.xml.in]: > /latest/

is not a good idea in general.

Please use `@OPENCV_MATHJAX_RELPATH@` here.

File: modules/java/generator/gen_java.py [hunk 2]
```diff
@@ -105,14 +106,134 @@ def __init__(self, type, decl, namespaces):
...
+            #logging.info('docstring: %s', doc)
+            if re.search("(@|\\\\)deprecated", doc):
                 self.annotation.append("@Deprecated")
 
+            if type == "class":
+                doc = doc.replace("@param ", "")
+
+            doc = re.sub(re.compile('\\\\f\\$(.*?)\\\\f\\$', re.DOTALL), '\\(' + r'\1' + '\\)', doc)
+            doc = re.sub(re.compile('\\\\f\\[(.*?)\\\\f\\]', re.DOTALL), '\\(' + r'\1' + '\\)', doc)
+            doc = re.sub(re.compile('\\\\f\\{(.*?)\\\\f\\}', re.DOTALL), '\\(' + r'\1' + '\\)', doc)
```
Comment (alalek) [inline modules/java/generator/gen_java.py]: Could you please move all this documentation code into a separate global function, like `sanitize_java_documentation_string()`?

File: modules/java/jar/build.xml.in (hunk 1 above)
Comment (asashour) [reply modules/java/jar/build.xml.in]: it seems the variable is not replaced.

Comment (alalek) [reply modules/java/jar/build.xml.in]: Right!
This variable is defined in `doc/CMakeLists.txt` which is processed after java module.

Please move this variable definition into the main [CMakeLists.txt](https://github.com/opencv/opencv/blob/3.4.6/CMakeLists.txt#L499).

**Pull request merged**